src/Databases/DatabasesCommon.h

```diff
@@ -52,6 +52,28 @@ class DatabaseWithOwnTablesBase : public IDatabase, protected WithContext
     void attachTableUnlocked(const String & table_name, const StoragePtr & table) TSA_REQUIRES(mutex);
     StoragePtr detachTableUnlocked(const String & table_name)  TSA_REQUIRES(mutex);
     StoragePtr getTableUnlocked(const String & table_name) const TSA_REQUIRES(mutex);
+
+    template <class DatabaseClass>
+    void renameTableImpl(ContextPtr local_context, const String & table_name, IDatabase & to_database, const String & to_table_name, bool exchange, bool dictionary);
+};
+
+template <bool lazy>
+class TableDataMapping
```

Comment:
I am not sure if this is the best solution, but now symlinks and data mappings are in one place.

I didn't add it to DatabaseWithOwnTablesBase because we would need to add `path_to_table_symlinks` to all of the constructors in DatabaseAtomic -> DatabaseOrdinary -> DatabaseOnDisk -> DatabaseWithOwnTablesBase, and together with `table_name_to_path` they will be hanging around in these classes doing nothing.

Saying that we also lose TSA checks, so maybe the second way is the best one.